onnxruntime/core/providers/cpu/ml/tree_ensemble_aggregator.h

```diff
@@ -0,0 +1,481 @@
+  const std::vector<OTYPE>* base_values_;
+  OTYPE origin_;
+  bool use_base_values_;
+
+ public:
+  inline _Aggregator(size_t n_trees,
+                     const int64_t& n_targets_or_classes,
+                     POST_EVAL_TRANSFORM post_transform,
+                     const std::vector<OTYPE>* base_values) : n_trees_(n_trees), n_targets_or_classes_(n_targets_or_classes), post_transform_(post_transform), base_values_(base_values) {
+    origin_ = base_values_->size() == 1 ? (*base_values_)[0] : 0.f;
+    use_base_values_ = base_values_->size() == static_cast<size_t>(n_targets_or_classes_);
+  }
+
+  // 1 output
+
+  inline void ProcessTreeNodePrediction1(OTYPE* /*predictions*/, TreeNodeElement<OTYPE>* /*rooOTYPE*/,
+                                         unsigned char* /*has_predictions*/) const {}
+
+  inline void MergePrediction1(OTYPE* /*predictions*/, unsigned char* /*has_predictions*/,
+                               OTYPE* /*predictions2*/, unsigned char* /*has_predictions2*/) const {}
+
+  inline void FinalizeScores1(OTYPE* Z, OTYPE& val,
+                              unsigned char& has_scores,
+                              int64_t*) const {
+    val = has_scores ? (val + origin_) : origin_;
+    *Z = this->post_transform_ == POST_EVAL_TRANSFORM::PROBIT ? static_cast<OTYPE>(ComputeProbit(static_cast<float>(val))) : val;
+  }
+
+  // N outputs
+
+  void ProcessTreeNodePrediction(OTYPE* /*predictions*/, TreeNodeElement<OTYPE>* /*rooOTYPE*/,
+                                 unsigned char* /*has_predictions*/) const {}
+
+  void MergePrediction(int64_t /*n*/, OTYPE* /*predictions*/, unsigned char* /*has_predictions*/,
+                       OTYPE* /*predictions2*/, unsigned char* /*has_predictions2*/) const {}
+
+  void FinalizeScores(std::vector<OTYPE>& scores,
+                      std::vector<unsigned char>& has_scores,
+                      OTYPE* Z, int add_second_class,
+                      int64_t*) const {
+    OTYPE val;
+    for (int64_t jt = 0; jt < n_targets_or_classes_; ++jt) {
+      val = use_base_values_ ? (*base_values_)[jt] : 0.f;
+      val += has_scores[jt] ? scores[jt] : 0;
+      scores[jt] = val;
+    }
+    write_scores(scores, post_transform_, Z, add_second_class);
+  }
+};
+
+/////////////
+// regression
+/////////////
+
+template <typename ITYPE, typename OTYPE>
+class _AggregatorSum : public _Aggregator<ITYPE, OTYPE> {
+ public:
+  inline _AggregatorSum(size_t n_trees,
+                        const int64_t& n_targets_or_classes,
+                        POST_EVAL_TRANSFORM post_transform,
+                        const std::vector<OTYPE>* base_values) : _Aggregator<ITYPE, OTYPE>(n_trees, n_targets_or_classes,
+                                                                                           post_transform, base_values) {}
+
+  // 1 output
+
+  inline void ProcessTreeNodePrediction1(OTYPE* predictions,
+                                         TreeNodeElement<OTYPE>* root,
```

Comment:
Lots of the methods here should be using references instead of pointers for both const and non-const arguments that cannot be nullptr.

Reply:
I changed root. I let the others as pointers. The functions suffixed by '1' usually handle one output but also 2 if the tree is a binary classifier (I need pointers in this case).